frontend/components/instruments/SpatialInstrumentInternal.js

```diff
@@ -0,0 +1,180 @@
+        minPointRadius=2,
+        maxPointRadius=5,
+        updateSize=(w, h)=>{},
+    }
+) => {
+    const audioContextRef = useRef(new AudioContext());
+
+    const instrumentDiv = useRef(null);
+    const [mouseX, setMouseX] = useState(null);
+    const [mouseY, setMouseY] = useState(null);
+
+    const [instrumentWidth, setInstrumentWidth] = useState(null);
+    const [instrumentHeight, setInstrumentHeight] = useState(null);
+
+    useEffect(() => {
+        const {width: w, height: h} = instrumentDiv.current.getBoundingClientRect();
+        setInstrumentWidth(w);
+        setInstrumentHeight(h);
+    }, [width, height, instrumentDiv]);
+
+    useEffect(() => {
+        updateSize(instrumentWidth, instrumentHeight);
+    }, [instrumentWidth, instrumentHeight]);
+
+    // Instrument properties
+    const maxDist = 100;
+    const halfRange = 40;
+    const visualizeRange = true;
+
+    function sqDist(x1, y1, x2, y2) {
+        return (x1 - x2)*(x1 - x2) + (y1 - y2)*(y1 - y2);
+    }
+
+    function mouseActive() {
+        return mouseX !== null && mouseY !== null;
+    }
+
+    function sampleShouldPlay(point) {
+        if (!mouseActive()) {
+            return false;
+        }
+        return sqDist(point.x, point.y, mouseX, mouseY) <= maxDist*maxDist;
+    }
+
+    function sampleVolume(point) {
+        if (!mouseActive()) {
+            return 0;
+        }
+        if (!sampleShouldPlay(point)) return 0;
+        return 100/(sqDist(point.x, point.y, mouseX, mouseY)/(halfRange*halfRange) + 1);
+    }
+
+    function sampleRadius(point) {
+        const vol = sampleVolume(point);
+        return minPointRadius + (vol/100) * (maxPointRadius - minPointRadius);
+    }
+
+    function handleMouseLeave() {
+        setMouseX(null);
+        setMouseY(null);
+    }
+
+    function handleMouseMove(e) {
+        const rect = instrumentDiv.current.getBoundingClientRect();
+
+        // update width + height if out of sync
+        if (instrumentWidth !== rect.width) {
+            setInstrumentWidth(rect.width);
+        }
+        if (instrumentHeight !== rect.height) {
+            setInstrumentHeight(rect.height);
+        }
+
+        const x = e.clientX - rect.left; // x position within the element.
+        const y = e.clientY - rect.top;  // y position within the element.
+        setMouseX(x);
+        setMouseY(y);
+    }
+
+    return (
+        <>
+            <svg
+                className={STYLES.spatialInstrument}
+                onMouseLeave={handleMouseLeave}
+                onMouseMove={handleMouseMove}
+                onMouseEnter={handleMouseMove}
+                ref={instrumentDiv}
+                width={width}
+                height={height}
+            >
+                <SvgRender
+                    width={instrumentWidth}
+                    height={instrumentHeight}
+                    x={mouseX}
+                    y={mouseY}
+                />
+                {soundPoints.map((soundPoint, i) => (
```

Comment:
I'd suggest factoring out the mouse circle thing into its own component and each one of these soundpoint display / sample player combos as a separate component, just for readability.